470 export section people

rmathew1011 opened the pull request:
# Description

 Added organization section export functionality for People view

 - Added endpoint: /individual/{id}/export?view=People in `IndividualExportController` to export view: people associated  with an organization

 - Implemented `getIndividualsData(id)` method to get related individuals by passing Organization id.

 - Added `streamIndividuals` for CSV exporter using to stream people data.

 - Added `export` property to `DisplayViews` to support custom export columns.

 - Updated display yml with desired export columns - `TAMU email`, `Orcid`, `Last name`, `First name`, `Preferred title`, `Position`, `Department` and `Scholars uri`.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing Procedures

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

qtamu committed:
Close and expand research areas

rmathew1011 committed:
WIP - export people for selected organization

rmathew1011 committed:
Removed the exportViews property from displaysection view

rmathew1011 committed:
WIP - added additional columns to export in display view

rmathew1011 committed:
WIP - arranged the order for column header in display view

rmathew1011 committed:
Merge remote-tracking branch 'origin/sprint12-staging' into test-export-section-people

rmathew1011 committed:
Removed debug statements

rmathew1011 committed:
Export button for people section works

rmathew1011 committed:
Update IndividualRepo.java

Cleaned up the code - removed debug lines.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/discovery/model/repo/IndividualRepo.java:
```diff
@@ -148,16 +150,51 @@ public List<Individual> findIndividualsByIds(List<String> ids) {
         }
     }
 
+    public List<Individual> findPeopleByOrganizationId(String orgId) {
```
Please update IndexDocumentRepo for additional interface methods. Please do not provide types in method or argument names.

tamu-sad-iii replied:
For collection class members, use an argument to provide the literal to get the collection from the individual.

I suggest something like this and do not include types not relevant to the encapsulation.

```
getIndividualsData(String id, String propertyName)
```

rmathew1011 replied:
updated the IndexDocumentRepo and refactored the method name as suggested above.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/discovery/model/repo/IndividualRepo.java:
```diff
@@ -148,16 +150,51 @@ public List<Individual> findIndividualsByIds(List<String> ids) {
+                return Collections.emptyList();
+            }
+            List<String> orgPeopleListIds = orgPeopleList.stream()
+                .map(s -> s.contains("::") ? s.split("::")[1] : s)
```
Please use a public static utility method for this.

rmathew1011 replied:
Utilized a static utility method for the above.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/export/controller/IndividualExportController.java:
```diff
@@ -103,6 +107,21 @@ public ResponseEntity<StreamingResponseBody> export(
+        @PathVariable String id,
+        @RequestParam(required = false, defaultValue = "People") String view,
+        @RequestParam(required = false, defaultValue = "csv") String type,
+        List<ExportArg> export
```
Please add @RequestBody to be explicit and improved readability.

rmathew1011 replied:
Used @RequestParam instead of @RequestBody

tamu-sad-iii replied:
Oh right. ExportArg have an argument resolver from request params. Sorry.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/export/controller/IndividualExportController.java:
```diff
@@ -103,6 +107,21 @@ public ResponseEntity<StreamingResponseBody> export(
         }
     }
 
+    @GetMapping("/individual/{id}/export/section")
```
This isn't according to API design yet. In order to request a section, you would require the display view and from the display view a section and from the section an export view. 

I recommend RESTful Collection/Resource API patterns.

```
/individual/{id}/section/{sectionId}/export
```

The display view can be looked up by individual type.

tamu-sad-iii replied:
I am not sure the integration of the export package semantics should include view package semantics in the API paths. So, this might not be ideal either. I think the existing export path is more than suitable and should be able to extend its functionality to support a given export view (export args) for a given individual and exporter type.

rmathew1011 replied:
replace with line: @GetMapping(value = "/individual/{id}/export", params = "view")

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/export/service/Exporter.java:
```diff
@@ -27,6 +27,13 @@ public default StreamingResponseBody streamIndividuals(Flux<Individual> cursor,
         ));
     }
 
+    public default StreamingResponseBody streamIndividuals(Individual individual, List<ExportArg> export) {
```
Please move this to the bottom with the other individual scoped streaming.

rmathew1011 replied:
Moved it to the bottom.

rmathew1011 committed:
- Incorporated PR suggestions
 - The discoveryViews people template had code in it that were not part of this feature branch.
 - Reverted the change in application.yml file as change was not part of the feature branch.

rmathew1011 committed:
Added a utility file - as part of PR suggestion

rmathew1011 committed:
removed commented code

rmathew1011 committed:
Extra space removed.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/discovery/model/repo/IndividualRepo.java:
```diff
@@ -148,16 +151,50 @@ public List<Individual> findIndividualsByIds(List<String> ids) {
+        try {
+            SolrQueryBuilder queryBuilder = new SolrQueryBuilder()
+                .withFilters(Arrays.asList(
+                    FilterArg.of("id", Optional.of(id), Optional.empty(), Optional.empty())
```
Please use an existing constant for "id".

rmathew1011 replied:
Utilized the existing `DiscoveryConstants.ID` for "id".

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/discovery/model/repo/IndividualRepo.java:
```diff
@@ -148,16 +151,50 @@ public List<Individual> findIndividualsByIds(List<String> ids) {
+
+            SolrDocument document = response.getResults().get(0);
+
+            List<String> orgPeopleList = (List<String>) document.getFieldValue("people");
```
Please make "people" a second argument of this method.

rmathew1011 replied:
the view "people" is now a second argument.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/discovery/model/repo/IndividualRepo.java:
```diff
@@ -210,6 +246,55 @@ public DiscoveryFacetAndHighlightPage<Individual> search(
         }
     }
 
+    public Flux<Individual>exportSection(QueryArg query, Sort sort, String view, String type, String id) {
```
Is this being called anywhere?

rmathew1011 replied:
Accidentally had this method pushed - deleted this method.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/export/controller/IndividualExportController.java:
```diff
@@ -103,6 +107,21 @@ public ResponseEntity<StreamingResponseBody> export(
     }
 
+    @GetMapping(value = "/individual/{id}/export", params = "view")
+    public ResponseEntity<StreamingResponseBody> exportSectionPeople(
```
Please rename this to not be specific to type People.

rmathew1011 replied:
Refactored to `exportSection`.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/export/utility/ExtractIdUtility.java:
```diff
@@ -0,0 +1,28 @@
+/**
+ * 
+ */
+public class ExtractIdUtility {
```
Is there an existing utility that can be used? I think there are other places the raw flatten Solr document for individuals are processed. I think this is avoidable if the nested documents are serialized from the flatten document.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/view/model/DisplaySectionView.java:
```diff
@@ -50,6 +50,13 @@ public class DisplaySectionView extends FieldView {
 
+    @JoinColumn(name = "export_view_id")
+    @OneToMany(cascade = CascadeType.ALL, fetch = FetchType.EAGER)
+    private List<ExportView> exportViews;
```
Is this a necessary change? Is it used anywhere?

rmathew1011 replied:
Not used anywhere- deleted this property.

rmathew1011 committed:
- Incoporated PR suggestions: https://github.com/TAMULib/scholars-discovery/pull/499
 - Code clean up.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/discovery/model/repo/IndividualRepo.java:
```diff
@@ -148,16 +151,50 @@ public List<Individual> findIndividualsByIds(List<String> ids) {
+        }
+
+        try {
+            SolrQueryBuilder queryBuilder = new SolrQueryBuilder()
```
Can this be [findById](https://github.com/TAMULib/scholars-discovery/blob/dc7822ab0ded74530dd13ebeff4004dab1485aa6/src/main/java/edu/tamu/scholars/middleware/discovery/model/repo/IndexDocumentRepo.java#L35) instead of using the solr client directly?

rmathew1011 replied:
Please allow me to create a note to address the above suggestion.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/export/controller/IndividualExportController.java:
```diff
@@ -103,6 +107,21 @@ public ResponseEntity<StreamingResponseBody> export(
     }
 
+    @GetMapping(value = "/individual/{id}/export", params = "view")
+    public ResponseEntity<StreamingResponseBody> exportSection(
```
Why not just `export`?

rmathew1011 replied:
There already exists a method export in this same file and so refactoring this method to `export`
resulted in the following error:
```Erasure of method export(String, String, String, List<ExportArg>) is the same as another method in type IndividualExportControllerJava```

tamu-sad-iii replied:
It is okay to have same method name with different arguments.

tamu-sad-iii replied:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.14.0:compile (default-compile) on project middleware: Compilation failure
[ERROR] /C:/Users/wwelling.AUTH/TL/Scholars/scholars-discovery/src/main/java/edu/tamu/scholars/middleware/export/controller/IndividualExportController.java:[111,50] name clash: export(java.lang.String,java.lang.String,java.lang.String,java.util.List<edu.tamu.scholars.middleware.export.argument.ExportArg>) and export(java.lang.String,java.lang.String,java.lang.String,java.util.List<java.lang.String>) have the same erasure
```

tamu-sad-iii reviewed (approved):
Works locally.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/view/model/DisplaySectionView.java:
```diff
@@ -50,6 +50,9 @@ public class DisplaySectionView extends FieldView {
     private List<DisplaySubsectionView> subsections;
 
+    @ElementCollection
+    private List<ExportField> export;
```
Can we make sure this corresponding change is reflected in scholars-angular core model?
